File: src/main/java/com/dingdangmaoup/resin/pura/resin/configuration/ResinConfigurationStrategy.java
package com.dingdangmaoup.resin.pura.resin.configuration;

import com.dingdangmaoup.resin.pura.ResinModel;
import com.dingdangmaoup.resin.pura.resin.ResinInstallation;
import com.dingdangmaoup.resin.pura.resin.WebApp;
import com.intellij.execution.ExecutionException;
import java.io.InputStream;
import org.jdom.Element;

public abstract class ResinConfigurationStrategy {
  private Element myElement;

  public void init(ResinModel serverModel, Element element) throws ExecutionException {
    myElement = element;
  }

  protected final Element getElement() {
    return myElement;
  }

  public void save() throws ExecutionException {

  }

  public abstract void setPort(int port);

  public abstract boolean deploy(WebApp webApp) throws ExecutionException;

  public abstract boolean undeploy(WebApp webApp) throws ExecutionException;

  public abstract InputStream getDefaultResinConfContent();

  public String getServerId() {
    return null;
  }

  public static ResinConfigurationStrategy getForInstallation(final ResinInstallation installation) {
    String verNumber = installation.getVersion().getVersionNumber();
    //Extract only first version number (to accept 2.x, 3.x, 3.1.13, ...)
    int resinVersion = Integer.parseInt(verNumber.substring(0, verNumber.indexOf('.')));
    int buildVersion = Integer.parseInt(verNumber.substring(2, 3));
    return switch (resinVersion) {
      case 2 -> new Resin2XConfigurationStrategy();
      case 3 ->
        //From resin 3.2.0 resin.conf is not longer valid... instead resin.xml
        switch (buildVersion) {
          case 0 -> new Resin3XConfigurationStrategy(installation);
          case 1 -> new Resin31ConfigurationStrategy(installation);
          default -> new ResinXmlConfigurationStrategy(installation);
        };
      case 4 ->
        //3.2 branch was renamed to 4.0 with some incompatible resin.xml changes
        new Resin4XmlConfigurationStrategy(installation);
      default -> new Resin3XConfigurationStrategy(installation);
    };
  }
}

File: src/main/kotlin/com/dingdangmaoup/resin/pura/resin/configuration/ResinConfigurationStrategy.kt
package com.dingdangmaoup.resin.pura.resin.configuration

import com.dingdangmaoup.resin.pura.ResinModel
import com.dingdangmaoup.resin.pura.resin.ResinInstallation
import com.dingdangmaoup.resin.pura.resin.WebApp
import com.intellij.execution.ExecutionException
import org.jdom.Element
import java.io.File
import java.io.InputStream

abstract class ResinConfigurationStrategy {
    private lateinit var myElement: Element
    private var mySourceConfig: File? = null

    @Throws(ExecutionException::class)
    internal fun init(serverModel: ResinModel, element: Element, sourceConfig: File) {
        mySourceConfig = sourceConfig.absoluteFile
        init(serverModel, element)
    }

    @Throws(ExecutionException::class)
    open fun init(serverModel: ResinModel, element: Element) {
        myElement = element
    }

    protected fun getElement(): Element = myElement

    protected fun getSourceConfigDirectory(): File? = mySourceConfig?.parentFile

    @Throws(ExecutionException::class)
    open fun save() {}

    abstract fun setPort(port: Int)

    @Throws(ExecutionException::class)
    abstract fun deploy(webApp: WebApp): Boolean

    @Throws(ExecutionException::class)
    abstract fun undeploy(webApp: WebApp): Boolean

    abstract fun getDefaultResinConfContent(): InputStream?

    open fun getServerId(): String? = null

    companion object {
        @JvmStatic
        fun getForInstallation(installation: ResinInstallation): ResinConfigurationStrategy {
            val versionParts = installation.getVersion().getVersionNumber().split('.')
            val resinVersion = versionParts.firstOrNull()?.toIntOrNull()
            val minorVersion = versionParts.getOrNull(1)?.toIntOrNull()
            return when (resinVersion) {
                2 -> Resin2XConfigurationStrategy()
                3 -> when (minorVersion) {
                    1 -> Resin31ConfigurationStrategy(installation)
                    null, 0 -> Resin3XConfigurationStrategy(installation)
                    else -> ResinXmlConfigurationStrategy(installation)
                }

                4 -> Resin4XmlConfigurationStrategy(installation)
                else -> Resin3XConfigurationStrategy(installation)
            }
        }
    }
}
